pep-0677.rst

```diff
@@ -184,33 +195,144 @@ This is shorter and requires fewer imports. It also has far less
-it was
-<https://github.com/pradeep90/annotation_collector#overall-stats-in-typeshed>`_
-the fifth most common complex type in typeshed, after ``Optional``,
-``Tuple``, ``Union``, and ``List``.
+A particularly common case where library authors leave off type information
+for callables is when defining decorators. Consider the following::
 
-Most of the other commonly used types have had their syntax improved
-via either PEP 604 or PEP 585. ``Callable`` is used heavily enough to
-similarly justify a more usable syntax.
 
-In this proposal, we chose to support all the existing semantics of
-``typing.Callable``, without adding support for new features. We took
-this decision after examining how frequently each feature might be
-used in existing typed and untyped open-source code. We determined
-that the vast majority of use cases are covered.
+    from typing import Any, Callable
+
+    def with_retries(
+        f: Callable[..., Any]
+    ) -> Callable[..., Any]:
+        def wrapper(retry_once, *args, **kwargs):
+            if retry_once:
+                try: return f(*args, **kwargs)
+                except Exception: pass
+            return f(*args, **kwargs)
+        return wrapper
+
+    @with_retries
+    def f(x: int) -> int:
+        return x
 
-We considered adding support for named, optional, and variadic
-arguments. However, we decided against including these features, as
-our analysis showed they are infrequently used. When they are really
-needed, it is possible to type these using `Callback Protocols
-<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.
 
-See the Rejected Alternatives section for more detailed discussion
-about omitted features.
+    f(y=10)  # oops - no type error!
+
+In the code above, it is clear that the decorator should produce a
+function whose signature is like that of the argument ``f`` other
+than an additional ``retry_once`` argument. But the use of ``...``
+prevents a type checker from seeing this and alerting a user that
+``f(y=10)`` is invalid.
+
+
+With PEP 612 it is possible to type decorators like this correctly
+as follows::
+
+    from typing import Any, Callable, Concatenate, ParamSpec, TypeVar
+
+    R = TypeVar("R")
+    P = ParamSpec("P")
+
+    def with_retries(
+        f: Callable[P, R]
+    ) -> Callable[Concatenate[bool, P] R]:
+        def wrapper(retry_once: bool, *args: P.args, **kwargs: P.kwargs):
+            ...
+        return wrapper
+
+    ...
+
+
+With our proposed syntax, the properly-typed decorator example becomes
+concise and the type representations are visually descriptive::
+
+    from typing import Any, ParamSpec, TypeVar
+
+    R = TypeVar("R")
+    P = ParamSpec("P")
+
+    def with_retries(
+        f: (**P) -> R
+    ) -> (bool, **P) -> R:
+        ...
+
+Comparing to Other Languages
+----------------------------
+
+Many popular programming languages use an arrow syntax similar
+to the one we are proposing here.
+
+TypeScript
+~~~~~~~~~~
+
```

Comment:
Could you mention how their function signatures look? That's one of the main differences from Python - they don't use an arrow for the return type of the function signature.

Reply:
Will do - I think I'll make the point in a separate subsubsection because that point actually applies to all of the non-ML-family examples (scala, kotline, typescript).

Reply:
Also, in the new subsubsection I'll call out our concern about whether the overloaded use of ``->`` presents readability problems.